app/javascript/utils/debounce.js

```diff
@@ -0,0 +1,18 @@
+window._ = {
+  debounce: function(func, wait, immediate) {
+    var timeout
+
+    return function() {
+      var context = this, args = arguments
+
+      var later = function() {
+        timeout = null
+        if (!immediate) func.apply(context, args)
+      }
+      var callNow = immediate && !timeout
+      clearTimeout(timeout)
+      timeout = setTimeout(later, wait)
+      if (callNow) func.apply(context, args)
+    }
+  }
+}
```

Comment:
@robacarp If I don't hang the function on window (like lodash does) then it's not visible inside the stimulus controller. Probably it's because I need to do an import? I figured: it'd be nice not to have to an import a collection of utility functions we build up over time so maybe I put them in the global namespace by copying this hack? Thoughts welcome.

Reply:
The es6/7 module js and import statements make it so this isn't necessary, but more importantly it provides a benefit too.

If for some reason you need to use two different debounce functions, you can do so and name them however you want in the module that needs them and not affect the rest of the application. When you declare it globally like this, if someone later loads lodash itself into the global namespace, it'll overwrite this and you end up with a strange bug that's difficult to find and hard to fix.

If you use it in your module you can do this and still use our code and some other code:

```javascript
import { debouce as ourDebouncer } from "../utils/debounce"
import { debounce as someOtherDebouncer } from "google_secret_sauce_debouncer"
```